### PR TITLE
[FLINK-30496][table-api] Introduce TableChange to represent MODIFY change

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Introduce TableChange to represent ALTER TABLE MODIFY syntax.*

## Brief change log

  - *Collect table changes during the sql to operation*
  - *Add the modify changes in the `TableChange`*

## Verifying this change

This change is already covered by existing tests, such as *SqlToOperationConverterTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 1]
```diff
@@ -151,12 +157,35 @@ public static Operation convertChangeColumn(
...
+                                .getResolvedSchema()
+                                .getColumn(oldName)
+                                .orElseThrow(
+                                        () ->
+                                                new ValidationException(
+                                                        "Failed to get old column: " + oldName)),
+                        Column.physical(newTableColumn.getName(), newTableColumn.getType())
+                                .withComment(
+                                        changeColumn
+                                                .getNewColumn()
```
Comment (LadyForest) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:178]: Add a util method?

```java
@Nullable
private static String getComment(SqlTableColumn column) {
    return column.getComment()
           .map(SqlCharStringLiteral.class::cast)
            .map(c -> c.getValueAs(String.class))
            .orElse(null);
}
```

**Review by LadyForest — CHANGES_REQUESTED** [review]
Thanks for the contribution! The PR looks good, whereas I have several questions.

1. Since the MODIFY operation is split into a list of changes, I would like to know whether the order matters and how the external catalog copes with them. For example, does API expose a requirement that it should be an atomic operation, or does the external catalog decide it?

2. The "oldColumn" and "originColumn" are interchangeably used across the changelog, and can we align them to improve the readability?

3. `SchemaConverter#updatePositionAndCollectColumnChange` and `OperationConverterUtils#buildColumnChange` are very similar. Do we have a plan to do some refactor work?

**Comment by fsk119** [comment]
Thanks for the review!

>Since the MODIFY operation is split into a list of changes, I would like to know whether the order matters and how the external catalog copes with them. For example, does API expose a requirement that it should be an atomic operation, or ? does the external catalog decide it?

Yes, the order matters. The catalog developer needs to make sure the execution in order and the execution is atomic. The logic may looks like(the idea is from the Spark JdbcTablecatalog#alterTable)

```

default void alterTable(
            ObjectPath tablePath,
            CatalogBaseTable newTable,
            List<TableChange> tableChanges,
            boolean ignoreIfNotExists)
            throws TableNotExistException, CatalogException {
    // make sure all table changes are supported.
    validate(tableChanges);
    // jdbc catalog can work like
    conn.setAutoCommit(false);
    Statement statement = conn.createStatement()
    try {
        statement.setQueryTimeout(options.queryTimeout);
        for (String sql = dialect.alterTable(tableName, changes, metaData.getDatabaseMajorVersion)) {
           statement.executeUpdate(sql)
        }
        conn.commit()
    } catch (Exception e) {
      if (conn != null) {
        conn.rollback();
      }
      throw e;
    } finally {
      statement.close();
      conn.setAutoCommit(true);
    }
}

``` 

> The "oldColumn" and "originColumn" are interchangeably used across the changelog, and can we align them to improve the readability?

Thanks for pointing it out. I agree with you we should align the term in the codes. I am prone to use the term "oldColumn" instead because we already have used this in the FLIP-273. WDYT?

> SchemaConverter#updatePositionAndCollectColumnChange and OperationConverterUtils#buildColumnChange are very similar. Do we have a plan to do some refactor work?

Yes. I think we can refactor this after all parts finish. WDYT?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java [hunk 2]
```diff
@@ -276,6 +409,359 @@ public String toString() {
...
+                    + "Column="
+                    + oldColumn
+                    + ", newPosition="
+                    + newPosition
+                    + '}';
+        }
+    }
+
+    /**
+     * A table change that modify the physical column data type.
```
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/TableChange.java:576]: No. If users want to support modifying the metadata column type, we should introduce another class named `ModifyMetadataColumn`, which may also involve the change of the metadata key.

**Comment by LadyForest** [comment]
> Thanks for the review!
> 
> > Since the MODIFY operation is split into a list of changes, I would like to know whether the order matters and how the external catalog copes with them. For example, does API expose a requirement that it should be an atomic operation, or ? does the external catalog decide it?
> 
> Yes, the order matters. The catalog developer needs to make sure the execution in order and the execution is atomic. The logic may looks like(the idea is from the Spark JdbcTablecatalog#alterTable)
> 
> ```
> 
> default void alterTable(
>             ObjectPath tablePath,
>             CatalogBaseTable newTable,
>             List<TableChange> tableChanges,
>             boolean ignoreIfNotExists)
>             throws TableNotExistException, CatalogException {
>     // make sure all table changes are supported.
>     validate(tableChanges);
>     // jdbc catalog can work like
>     conn.setAutoCommit(false);
>     Statement statement = conn.createStatement()
>     try {
>         statement.setQueryTimeout(options.queryTimeout);
>         for (String sql = dialect.alterTable(tableName, changes, metaData.getDatabaseMajorVersion)) {
>            statement.executeUpdate(sql)
>         }
>         conn.commit()
>     } catch (Exception e) {
>       if (conn != null) {
>         conn.rollback();
>       }
>       throw e;
>     } finally {
>       statement.close();
>       conn.setAutoCommit(true);
>     }
> }
> ```
> 
> > The "oldColumn" and "originColumn" are interchangeably used across the changelog, and can we align them to improve the readability?
> 
> Thanks for pointing it out. I agree with you we should align the term in the codes. I am prone to use the term "oldColumn" instead because we already have used this in the FLIP-273. WDYT?
> 
> > SchemaConverter#updatePositionAndCollectColumnChange and OperationConverterUtils#buildColumnChange are very similar. Do we have a plan to do some refactor work?
> 
> Yes. I think we can refactor this after all parts finish. WDYT?

Thanks for the detailed explanation. +1 for refactoring after all the subtasks finished

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java [hunk 3]
```diff
@@ -698,7 +782,7 @@ private AlterSchemaStrategy computeAlterSchemaStrategy(SqlAlterTableSchema alter
                         alterTableSchema.getClass().getCanonicalName()));
     }
 
-    private static <T> T unwrap(Optional<T> value) {
+    private <T> T unwrap(Optional<T> value) {
```
[the remark this answers is not in the document]
Comment (fsk119) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java:785]: ModifySchemaConverter and AddSchemaConverter are both static classes and can only use static methods in the parent class. But now I change them to class-level inner class, so I think we can remove the static keyword now. 

BTW, I think we should narrow the scope of the visibility. With the `static` keyword, the visibility is much larger. WDYT?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java [hunk 4]
```diff
@@ -685,18 +731,18 @@ String getComment(SqlTableColumn column) {
     // --------------------------------------------------------------------------------------------
 
     private void validateColumnName(
-            String originColumnName,
+            String oldColumnName,
             String newColumnName,
-            ResolvedSchema originSchemas,
+            ResolvedSchema oldSchemas,
```
Comment (LadyForest) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java:736]: Nit: `oldSchemas` -> `oldSchema` ?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java [hunk 5]
```diff
@@ -685,18 +731,18 @@ String getComment(SqlTableColumn column) {
...
             List<String> partitionKeys) {
         validateColumnName(
-                originColumnName,
-                originSchemas,
+                oldColumnName,
+                oldSchemas,
```
Comment (LadyForest) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterSchemaConverter.java:740]: Nit: `oldSchemas` -> `oldSchema` ?

**Review by LadyForest — COMMENTED** [review]
Hi, @fsk119 thanks for the update! The PR looks pretty good and I only left two minor comments.

**Review by LadyForest — APPROVED** [review]
Thanks for the update, LGTM 👍

**Comment by fsk119** [comment]
The failed test is because of FLINK-30257. Merging...

**Pull request closed without merging**